server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java

```diff
@@ -0,0 +1,133 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.joining;
+
+import java.io.InputStream;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BlobStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.annotations.VisibleForTesting;
+
+public class JoiningBlobStore implements BlobStore {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JoiningBlobStore.class);
+
+    private final BlobStore primaryBlobStore;
+    private final BlobStore secondaryBlobStore;
+
+    @VisibleForTesting
+    JoiningBlobStore(BlobStore primaryBlobStore, BlobStore secondaryBlobStore) {
+        this.primaryBlobStore = primaryBlobStore;
+        this.secondaryBlobStore = secondaryBlobStore;
+    }
+
+    @Override
+    public CompletableFuture<BlobId> save(byte[] data) {
+        try {
+            return saveToPrimaryFallbackIfFails(
+                primaryBlobStore.save(data),
+                () -> secondaryBlobStore.save(data));
+        } catch (Exception e) {
+            LOGGER.error("exception directly happens while saving bytes data, fall back to secondary blob store", e);
+            return secondaryBlobStore.save(data);
```

Comment:
I think this operation has already been done, no?
Then you should throw here.

Reply:
No not necessarlily, it depends at which point you fail.

A failure can be **an exception beuing thrown** or **an exception during the future execution**. Both needs to be addressed.

Reply:
yeah, `secondaryBlobStore.save(data)` in try block is passed as a supplier. I don't execute it util first completableFuture completed exceptionally. This one is for addressing execption happens directly in method call, not in while executing future.